pilot/pkg/proxy/envoy/gateway.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package envoy
+
+import (
+	"strings"
+
+	"reflect"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	routing "istio.io/api/routing/v1alpha2"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/log"
+)
+
+func buildGatewayListeners(mesh *meshconfig.MeshConfig, instances []*model.ServiceInstance, config model.IstioConfigStore, ingress model.Node) Listeners {
+	gateways, err := config.List(model.Gateway.Type, model.NamespaceAll)
+	if err != nil {
+		log.Warnf("error listing gateways: %v", err)
+		return Listeners{}
+	} else if len(gateways) == 0 {
+		log.Debugf("listing gateways in namespace %q, got none", model.NamespaceAll)
+		return Listeners{}
+	}
+
+	gateway := gateways[0].Spec.(*routing.Gateway)
+	for _, spec := range gateways[1:] {
+		model.MergeGateways(gateway, spec.Spec.(*routing.Gateway))
+	}
+
+	listeners := make(Listeners, 0, len(gateway.Servers))
+	for _, server := range gateway.Servers {
+		opts := httpListenerOpt{
+			mesh:             mesh,
+			node:             ingress,
+			instances:        instances,
+			routeConfig:      nil,
+			ip:               WildcardAddress,
+			port:             int(server.Port.Number),
+			rds:              server.Port.Name,
+			useRemoteAddress: true,
+			direction:        IngressTraceOperation,
+			outboundListener: false,
+			store:            config,
+		}
+
+		switch strings.ToUpper(server.Port.Protocol) {
+		case "HTTP":
+			listener := buildHTTPListener(opts)
+
+			listeners = append(listeners, listener)
+		case "HTTPS", "HTTP2":
+			listener := buildHTTPListener(opts)
+
+			if server.Tls != nil {
+				listener.SSLContext = tlsToSSLContext(server.Tls, server.Port.Protocol)
+			}
+
+			listeners = append(listeners, listener)
+		case "TCP":
+			log.Warnf("TCP protocol support for Gateways is not yet implemented")
+		case "GRPC":
+			log.Warnf("gRPC protocol support for Gateways is not yet implemented")
+		case "MONGO":
+			log.Warnf("Mongo protocol support for Gateways is not yet implemented")
+		default:
+			log.Warnf("Gateway with invalid protocol: %q; %v", server.Port.Protocol, server)
+		}
+	}
+
+	return dedupeListeners(listeners)
```

Comment:
we have this normalize() function for listeners, clusters, routes, etc. that essentially does dedup